File: maze-dfs/Cell.js
function Cell(i, j) {
  this.i = i;
  this.j = j;
  this.visited = false;
  this.walls = {top: true, right: true, bottom: true, left: true}
  
  this.show = function() {
    const coordX = this.i * view;
    const coordY = this.j * view;
    
    stroke(255);
    
    // draw cell walls if they exist
    if (this.walls.top) {
      line(coordX, coordY, coordX + view, coordY);
    }
    
    if (this.walls.right) {
      line(coordX + view, coordY, coordX + view, coordY + view);
    }
    
    if (this.walls.bottom) {
      line(coordX + view, coordY + view, coordX, coordY + view);
    }
    
    if (this.walls.left) {
      line(coordX, coordY + view, coordX, coordY);
    }
    
    if (this.visited) {
      noStroke();
      fill(0);
      rect(coordX, coordY, view, view);
    }
    
  }
  
  this.highlight = function() {
    noStroke();
    fill(255);
    rect(this.i * view, this.j * view, view, view, 20);
  }
  
  this.getNeighbor = function() {
    const neighbors = [];
    
    const top = cells[getNeighborIndex(i - 1, j)];
    const right = cells[getNeighborIndex(i, j + 1)];
    const bottom = cells[getNeighborIndex(i + 1, j)];
    const left = cells[getNeighborIndex(i, j - 1)];
    
    if (top && !top.visited) neighbors.push(top);
    if (right && !right.visited) neighbors.push(right);
    if (bottom && !bottom.visited) neighbors.push(bottom);
    if (left && !left.visited) neighbors.push(left);
    
    if (neighbors.length > 0) {
      const randomNeighbor = floor(random(0, neighbors.length));
    
      return neighbors[randomNeighbor];
    }
  }
}

function getNeighborIndex(i, j) {
  // edge cases (no neighbor)
  if (i < 0 || j < 0 || i > rows - 1 || j > cols - 1) {
    return -1;
  }
    
  return j + i * cols; 
}

File: sketches/dfs_maze/Cell.py
class Cell(object):
    def __init__(self, i, j):
        self.i = i
        self.j = j
        self.visited = False
        self.walls = {"top": True, "right": True, "bottom": True, "left": True}
        
        
    def show(self, view):
        x = self.i * view
        y = self.j * view
        
        stroke(255)
        
        # draw cell walls if they exist
        if self.walls["top"]:
            line(x, y, x + view, y)
        if self.walls["right"]:
            line(x + view, y, x + view, y + view)
        if self.walls["bottom"]:
            line(x + view, y + view, x, y + view)
        if self.walls["left"]:
            line(x, y + view, x, y)
            
        if self.visited:
            noStroke()
            fill(0)
            rect(x, y, view, view)
    
    
    def highlight(self, view):
        noStroke()
        fill(255)
        rect(self.i * view, self.j * view, view, view, 20)
        
        
    @staticmethod
    def get_neighbor_index(i, j, rows, cols):
        # edge cases (no neighbor)
        if i < 0 or j < 0 or i > rows - 1 or j > cols - 1:
            return -1
        return j + i * cols
    
    
    def get_neighbor(self, cells, rows, cols):
        neighbors = []
        
        top = cells[self.get_neighbor_index(self.i - 1, self.j, rows, cols)]
        right = cells[self.get_neighbor_index(self.i, self.j + 1, rows, cols)]
        bottom = cells[self.get_neighbor_index(self.i + 1, self.j, rows, cols)]
        left = cells[self.get_neighbor_index(self.i, self.j - 1, rows, cols)]
        
        if top and not top.visited:
            neighbors.append(top)
        if right and not right.visited:
            neighbors.append(right)
        if bottom and not bottom.visited:
            neighbors.append(bottom)
        if left and not left.visited:
            neighbors.append(left)
        
        if neighbors:
            random_index = floor(random(0, len(neighbors)))
            
            return neighbors[random_index]
        
